streams/src/test/java/org/apache/kafka/streams/integration/KStreamTransformIntegrationTest.java

```diff
@@ -0,0 +1,306 @@
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Properties;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+@Category({IntegrationTest.class})
+public class KStreamTransformIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER =
+        new EmbeddedKafkaCluster(NUM_BROKERS);
+
+    private static volatile int testNo = 0;
+    private final MockTime mockTime = CLUSTER.time;
+    private StreamsBuilder builder;
+    private Properties streamsConfiguration;
+    private KafkaStreams kafkaStreams;
+    private String streamOneInput;
+    private String userSessionsStream = "user-sessions";
+    private String outputTopic;
+    private KStream<Integer, Integer> stream;
+    private StoreBuilder<KeyValueStore<Integer, Integer>> storeBuilder;
+
+    @Before
+    public void before() throws InterruptedException {
+        testNo++;
+        builder = new StreamsBuilder();
+        createTopics();
+        streamsConfiguration = new Properties();
+        final String applicationId = "kgrouped-stream-test-" + testNo;
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration
+            .put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(IntegrationTestUtils.INTERNAL_LEAVE_GROUP_ON_CLOSE, true);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        final StoreBuilder<KeyValueStore<Integer, Integer>> keyValueStoreBuilder =
+                Stores.keyValueStoreBuilder(Stores.persistentKeyValueStore("myTransformState"),
+                                            Serdes.Integer(),
+                                            Serdes.Integer());
+        builder.addStateStore(keyValueStoreBuilder);
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void shouldFlatTransform() throws Exception {
+
+        stream = builder.stream(streamOneInput, Consumed.with(Serdes.Integer(), Serdes.Integer()));
+        stream
+            .flatTransform(new TransformerSupplier<Integer, Integer, List<KeyValue<Integer, Integer>>>() {
```

Comment:
Nit: We typically use SAM conversion whenever possible just to keep the testing code compact.

Reply:
See my other comment about SAM conversion.